crates/compiler/load_internal/src/file.rs

```diff
@@ -4413,17 +4251,41 @@ fn send_header_two<'a>(
                 debug_assert!(!scope.contains_key(&ident));
                 scope.insert(ident, (symbol, entry.region));
             }
-        }
 
-        let ident_ids = ident_ids_by_module.get_mut(&home).unwrap();
+            // Generate IdentIds entries for all values this module exposes.
+            // This way, when we encounter them in Defs later, they already
+            // have an IdentIds entry.
+            //
+            // We must *not* add them to scope yet, or else the Defs will
+            // incorrectly think they're shadowing them!
+            for loc_provides in provides.iter() {
+                // Use get_or_insert here because the ident_ids may already
+                // created an IdentId for this, when it was imported exposed
+                // in a dependent module.
+                //
+                // For example, if module A has [B.{ foo }], then
+                // when we get here for B, `foo` will already have
+                // an IdentId. We must reuse that!
+                let ident_id = ident_ids.get_or_insert(loc_provides.value.as_str());
+                let symbol = Symbol::new(home, ident_id);
+
+                provided.push(symbol);
+            }
+
+            ident_ids.clone()
+        } else {
+            let ident_ids = ident_ids_by_module.get_mut(&home).unwrap();
+
+            ident_ids.clone()
+        };
 
         // Generate IdentIds entries for all values this module exposes.
         // This way, when we encounter them in Defs later, they already
         // have an IdentIds entry.
         //
         // We must *not* add them to scope yet, or else the Defs will
         // incorrectly think they're shadowing them!
-        for loc_exposed in provides.iter() {
+        for loc_exposed in exposes.iter() {
```

Comment:
The `provides` one happens earlier now, and only with platforms. (Platforms now also have `exposes` processed, which used to get ignored. Oops!)